File: rSVParser/src/packages.rs
//-----------------------------------------------------------------------------
// File: src/packages.rs
// Author: David Lenfesty
// Copyright (c) 2020. Eidetic Communications Inc.
// All rights reserved.
// Licensed under the BSD 3-Clause license.
// This license message must appear in all versions of this code including
// modified versions.
//----------------------------------------------------------------------------
use std::vec::Vec;

use sv_parser::{unwrap_node, SyntaxTree, RefNode};

use crate::get_identifier;
use crate::out::SvPackage;

/// Parses full file tree for package information.
pub fn parse_tree(tree: &SyntaxTree, path: &String) -> Vec<SvPackage> {
    let mut packages = Vec::new();

    for node in tree {
        match node {
            RefNode::PackageDeclaration(x) => {
                // XXX this could probably be a macro
                let name = unwrap_node!(x, PackageIdentifier).unwrap();
                let name = get_identifier(name).unwrap();
                let name = String::from(tree.get_str(&name).unwrap().trim());

                packages.push(SvPackage {
                    name: name.clone(),
                    endname: name,
                    include: false,
                    path: path.clone(),
                });
            },
            _ => (),
        }
    }

    return packages;
}

File: rSVParser/sv_hierarchy/packages.py
import pathlib
from typing import Tuple, List
from dataclasses import dataclass
from pathlib import Path

from py_sv_parser import SyntaxTree, unwrap_node


@dataclass
class SvPackage:
    name: str
    endname: str
    ports: List[Tuple]
    include: bool
    path: Path

def parse_tree(tree: SyntaxTree, path: Path):
    """Parses file tree for package information"""
    packages = []

    for node in tree:
        if node.type_name == "PackageDeclaration":
            name = unwrap_node(node, ["PackageIdentifier"])
            name = tree.get_str(name)

            packages.append(SvPackage(
                name, name, False, path
            ))

    return packages